Described table compatibility and use in place of merging to verify that...

gkellogg opened the pull request:
... embedded metadata is compatible with found metadata.

Remove merge, and merge descriptions.
Update description of creating embedded metadata to create empty column descriptions if header row count is zero.
This fixes #511.

@iherman @JeniT this could use a thorough review.

gkellogg committed:
Described table compatibility and use in place of merging to verify that embedded metadata is compatible with found metadata.
Remove merge, and merge descriptions.
Update description of creating embedded metadata to create empty column descriptions if header row count is zero.
This fixes #511.

iherman commented on syntax/index.html:
```diff
@@ -379,28 +379,18 @@
       <p>
-        Processors MUST always use overriding metadata and embedded metadata. If the overriding metadata does not originate from a complete metadata file itself (for example, if it is generated from command-line options when invoking the implementation), processors MUST attempt to locate the first metadata document from the <code>Link</code> header, the file-specific metadata, or the directory-sp ...
+        Processors MUST use the first metadata found for processing a <a>tabular data file</a> by using overriding metadata, if provided. Otherwise processors MUST attempt to locate the first metadata document from the <code>Link</code> header, the file-specific metadata, or the directory-specific metadata. If no metadata is supplied or found, processors MUST use embedded metadata. Processors MUS ...
       </p>
```
I do not understand the statement

> Processors MUST verify that the metadata is <a href="http://w3c.github.io/csvw/metadata/#dfn-table-compatibility" class="externalDFN">compatible</a> with the tabular data file as defined in [[!tabular-metadata]]." 

The section describes how to get to a specific metadata; compatibility is among two metadata.

Maybe the intention was to compare the _embedded_ metadata with whatever is found, to ensure the title's compatibility.

gkellogg replied:
Right, this says how to find the metadata, starting with overriding metadata, looking at link-, file-, and directory- metadata. Failing that, it uses embedded metadata. Whatever metadata is chosen, we ensure that it is compatible with the embedded metadata.

gkellogg replied:
It may be that the last sentence about verifying compatibility should be removed from this section, as it is described again in 6.1

iherman replied:
It doesn't harm to repeat; maybe something like

> Processors MUST verify that the metadata is compatible with the embedded metadata (if any).

Or something like that.

gkellogg replied:
The procedure a.ways results in embedded metadata, even if _header_ is false. It just isn't used to merge.

iherman commented on syntax/index.html:
```diff
@@ -587,15 +578,16 @@
                 <p>Parse the tabular data file, using <code>DD</code> as a guide, to create a basic tabular data model (<code>T</code>) and extract <a>embedded metadata</a> (<code>EM</code>), for example from the <a>header line</a>.</p>
                 <p class="note">This specification provides a non-normative definition for parsing CSV-based files, including the extraction of <a>embedded metadata</a>, in <a href="#parsing" class="sectionRef"></a>. This specification does not define any syntax for embedded metadata beyond this; whatever syntax is used, it's assumed that metadata can be mapped to the vocabulary defined in [[!tab ...
```
I am not sure what happens here, I do not see the whole process: we have a metadata file and then we extract the embedded metadata. Then what? There is no merge, so would the embedded metadata override what was used initially?

gkellogg replied:
Embedded metadata is used to determine table compatibility between `TM` and `EM`. Compatibility is defined in the metadata document. Embedded metadata does not override anything; we determine that it is compatible, and then process the CSV. Note that DD is modified, and `lang` may be updated in `TM`.

iherman replied:
Ah, yep, I found it below, sorry for the noise. 

However... then there should be some statement to signal what a processor should do if `EM` and `TM` are _not_ compatible. I presume that is an error condition, meaning the processing on that particular table (or the whole table group?) should stop, or something like that. I believe this belongs to this section.

gkellogg replied:
Yes, I'll add that. It is an error, similar to if a merge had failed.

iherman commented on metadata/index.html:
```diff
@@ -2177,19 +1936,11 @@
     <section class="appendix" id="changes">
-      <h2>Changes since the working draft of 08 January 2015</h2>
-      <p>The document has undergone substantial changes since the last working draft. Below are some of the changes made:</p>
+      <h2>Changes since the working draft of 16 April 2015</h2>
```
I think it is worth keeping the previous set of changes and add these as separate entries. If somebody wants to follow up history, it is more convenient to find everything at one place (a comment we did get when we published RDFa Edited Rec...)

gkellogg replied:
I'll put it back; I considered this, but it meant creating sub-sections in a changes appendix, and strictly speaking, this is describing changes from the previous WD, but there's value in keeping the history in the document, at least through LCCR.

gkellogg committed:
Add back previous change summary in a sub-section.
Update header block with correct previous version information.

gkellogg committed:
Indicate that an error should be raised if embedded metadata is not compatible with the specified or found table metadata, both in processing instructions, and for Validator requirements.

gkellogg commented:
@iherman see if the latest changes make it more clear.

iherman commented on syntax/index.html:
```diff
@@ -554,28 +544,29 @@
-          <li>Retrieve the metadata file yielding the metadata <code>UM<sub>M</sub></code> (which is treated as overriding metadata, see <a href="#overriding-metadata" class="sectionRef"></a>).</li>
-          <li>For each <a>table</a> in <code>UM<sub>M</sub></code> in order, create one or more <a title="annotated table">annotated tables</a>:
+          <li>Retrieve the metadata file yielding the metadata <code>UM</code> (which is treated as overriding metadata, see <a href="#overriding-metadata" class="sectionRef"></a>).</li>
+          <li>Normalize <code>UM</code> using the process defined in <cite><a href="http://w3c.github.io/csvw/metadata/#normalization">Normalization</a></cite> in [[!tabular-metadata]], coercing <code>UM</code> into a <a href="http://w3c.github.io/csvw/metadata/#dfn-table-group-description" class="externalDFN">table group description</a>, if necessary.</li>
```
(Nit picking...) Shouldn't it be 'normalization' instead of 'Normalization'?

gkellogg replied:
The name of the section this references is "Normalization", why would you think it should be lower-case?

iherman replied:
Do we capitalize each notion in the text? I am just trying to have consistency...

> On 26 Apr 2015, at 17:43, Gregg Kellogg notifications@github.com wrote:
> 
> In syntax/index.html:
> 
> > ```
> >      </ol>
> >      <p>If the process starts with a metadata file:</p>
> >      <ol class="algorithm">
> > ```
> > -          <li>Retrieve the metadata file yielding the metadata <code>UM<sub>M</sub></code> (which is treated as overriding metadata, see <a href="#overriding-metadata" class="sectionRef"></a>).</li>
> > -          <li>For each <a>table</a> in <code>UM<sub>M</sub></code> in order, create one or more <a title="annotated table">annotated tables</a>:
> > -          <li>Retrieve the metadata file yielding the metadata <code>UM</code> (which is treated as overriding metadata, see <a href="#overriding-metadata" class="sectionRef"></a>).</li>
> > -          <li>Normalize <code>UM</code> using the process defined in <cite><a href="http://w3c.github.io/csvw/metadata/#normalization">Normalization</a></cite> in [[!tabular-metadata]], coercing <code>UM</code> into a <a href="http://w3c.github.io/csvw/metadata/#dfn-table-group-description" class="externalDFN">table group description</a>, if necessary.</li>
> >   The name of the section this references is "Normalization", why would you think it should be lower-case?
> 
> —
> Reply to this email directly or view it on GitHub.

gkellogg replied:
This is a section reference, not a _notion_. IMO, the property descriptions should have probably been dfns rather than simple anchors, to make this clear; in this case, it's wrapped in a _cite_ element, which is my convention, at least, when referencing a section in another document, along with the actual document reference.

iherman commented on metadata/index.html:
```diff
@@ -880,6 +852,10 @@
+        <section>
+          <h4>Schema Compatibility</h4>
+          <p>Two Schemas are <dfn title="schema compatible">compatible</dfn> if they have the same number of non-<a href="#column-virtual"><code>virtual</code></a> columns, and all non-<a href="#column-virtual"><code>virtual</code></a> columns are compatible with each other. Columns at the same index are compatible if there is a non-empty case-insensitive intersection between the percent-decoded  ...
+        </section>
```
I am not sure of the he 

> Columns at the same index are compatible if there is a non-empty case-insensitive intersection between the percent-decoded <a href="#column-name"><code>name</code></a> and <a href="#column-titles"><code>titles</code></a> values.

Text. It may be read as if we'd require to compare names in one metadata with titles _in the other_. 

> Columns at the same index are compatible if there is a non-empty case-insensitive intersection between the percent-decoded <a href="#column-name"><code>name</code></a> and <a href="#column-titles"><code>titles</code></a> values of the respective column descriptions.

gkellogg replied:
That is, in fact, what we're doing. This allows metadata to define `name`, and compare with the equivalent `titles` from embedded metadata, for example. This text is lifted directly from the previous merge language on schema. The mechanism is to create an array containing the percent-decoded name, and all string values from `titles`, and see if there's an intersection.

Without comparing `name` against `titles`, every metadata column needs both `name` and `titles` annotations, which is overly redundant.

gkellogg replied:
We do need to add a case when there is an empty column description, as would be created in embedded metadata if `headers` is false. In this case, there are no titles, but we have a column count, and these columns should match any column. See, for example test019 in http://w3c.github.io/csvw/tests/.

iherman replied:
@gkellogg my only change was to emphasize that the comparison is among two colum descriptions. This was clear in the original text that you reused but, imho, not here...

iherman commented:
My comments are minor; I think we should merge these asap, so that we can notify implementers about the changes. Their implementation job will become simpler, let us avoid them unnecessary trouble:-)

gkellogg committed:
Add caveat that a column is compatible if it has neither name nor titles properties.

gkellogg committed:
Change schema compatibility to use defined `column description` rather than _columns_, which may be more vague.

gkellogg commented:
@iherman your comment about comparisons being between two column descriptions seems to have gotten lost. I did update the text again, see if that helps. Note that _schema compatility_ talks about `column descriptions` at the same index being compatible, without defining `column compatibility` separately; this seemed like overkill to me. Also, the fact that `column descriptions` are values of the `columns` property is implied and not make explicit.

Please suggest some clearer wording if you think it is still too vague.

iherman commented:
I would say merge it and close the current issue. I do not want to do the merge, though; @gkellogg or @JeniT should decide...

iherman commented on syntax/index.html:
```diff
@@ -587,15 +578,16 @@
-              <li>Create the merged metadata <code>M = merge(UM<sub>M</sub>, EM)</code> using the <a href="http://w3c.github.io/csvw/metadata/#dfn-merged" class="externalDFN">merge algorithm</a> defined in [[!tabular-metadata]].</li>
-              <li>Use the metadata <code>M</code> to add annotations to the tabular data model <code>T</code> as described in <a href="http://w3c.github.io/csvw/metadata/#annotating-tables">Section 2 Annotating Tables</a> in [[!tabular-metadata]].</li>
+              <li>If a <code>Content-Language</code> HTTP header was found when retrieving the tabular data file, and the value provides a single language, set the <a href="http://w3c.github.io/csvw/metadata/#cell-language" class="externalDFN">lang</a> inherited property to this value in <code>TM</code>, unless <code>TM</code> already has a <a href="http://w3c.github.io/csvw/metadata/#cell-langua ...
+              <li>Verify that <code>TM</code> is compatible <code>EM</code> using the procedure defined in <cite><a href="http://w3c.github.io/csvw/metadata/#table-compatibility">Table Compatibility</a> </cite> in [[!tabular-metadata]]; if <code>TM</code> is not compatible <code>EM</code> processors MUST raise an error.</li>
```
I guess it should be "is compatible with <code>EM</code>" (the 'with' is missing. Same at the end of the sentence.

gkellogg commented:
The normal practice @JeniT and I use is to have each other close the respective pull requests; this ensures review,by the other primary editor. jeni's schedule seems to be to do these before the call. If not done by the meeting, we can close it then.

gkellogg committed:
Missing "with". Thanks @iherman.

Commit:
compatibility of descriptions, only tested when validating

changed to table description compatibility, not table compatibility
added table group description compatibility
only validators need to check compatibility
moved embedded metadata section down so order makes more sense

JeniT commented:
@gkellogg please check you're happy with the changes I made in https://github.com/w3c/csvw/commit/90051c9e8d99839774d0d4cfc5b0f86bbead566d

gkellogg committed:
Table Description Compatibility uses `normalized url`.
